flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/lineage/LineageUtil.java

```diff
@@ -0,0 +1,118 @@
+public class LineageUtil {
+
+    private static final String KAFKA_DATASET_PREFIX = "kafka://";
+    private static final String COMMA = ",";
+    private static final String SEMICOLON = ";";
+
+    public static LineageDataset datasetOf(String namespace, KafkaDatasetFacet kafkaDatasetFacet) {
+        return datasetOf(namespace, kafkaDatasetFacet, Collections.emptyList());
+    }
+
+    public static LineageDataset datasetOf(
+            String namespace, KafkaDatasetFacet kafkaDatasetFacet, TypeDatasetFacet typeFacet) {
+        return datasetOf(namespace, kafkaDatasetFacet, Collections.singletonList(typeFacet));
+    }
+
+    private static LineageDataset datasetOf(
+            String namespace,
+            KafkaDatasetFacet kafkaDatasetFacet,
+            List<LineageDatasetFacet> facets) {
+        return new LineageDataset() {
+            @Override
+            public String name() {
+                return kafkaDatasetFacet.getTopicIdentifier().toLineageName();
+            }
+
+            @Override
+            public String namespace() {
+                return namespace;
+            }
+
+            @Override
+            public Map<String, LineageDatasetFacet> facets() {
+                Map<String, LineageDatasetFacet> facetMap = new HashMap<>();
+                facetMap.put(DefaultKafkaDatasetFacet.KAFKA_FACET_NAME, kafkaDatasetFacet);
+                facetMap.putAll(
+                        facets.stream()
+                                .collect(
+                                        Collectors.toMap(LineageDatasetFacet::name, item -> item)));
+                return facetMap;
+            }
+        };
+    }
+
+    public static String namespaceOf(Properties properties) {
+        String bootstrapServers = properties.getProperty("bootstrap.servers");
+
+        if (bootstrapServers == null) {
+            return KAFKA_DATASET_PREFIX;
+        }
+
+        if (bootstrapServers.contains(COMMA)) {
+            bootstrapServers = bootstrapServers.split(COMMA)[0];
+        } else if (bootstrapServers.contains(SEMICOLON)) {
+            bootstrapServers = bootstrapServers.split(SEMICOLON)[0];
+        }
+
+        return String.format(KAFKA_DATASET_PREFIX + bootstrapServers);
+    }
+
+    public static SourceLineageVertex sourceLineageVertexOf(Collection<LineageDataset> datasets) {
+        return new SourceLineageVertex() {
+            @Override
+            public Boundedness boundedness() {
+                return Boundedness.CONTINUOUS_UNBOUNDED;
+            }
+
+            @Override
+            public List<LineageDataset> datasets() {
+                return datasets.stream().collect(Collectors.toList());
+            }
+        };
+    }
+
+    public static LineageVertex lineageVertexOf(Collection<LineageDataset> datasets) {
+        return new LineageVertex() {
+            @Override
+            public List<LineageDataset> datasets() {
+                return datasets.stream().collect(Collectors.toList());
```

Comment:
Can we not use the `List.copyOf(datasets);` as per Arvid's suggested change here also.

Reply:
Merged version is not using `List.copyOf`. We agreed on that in offline discussion.